scribble-lib/scriblib/gui-eval.rkt

```diff
@@ -68,61 +66,61 @@
-              (let ([v (read log-file)])
-                (if (eof-object? v)
-                    (error "expression not in log file")
-                    (let ([v (deserialize v)])
-                      (if (equal? v (if (syntax? expr)
-                                        (syntax->datum expr)
-                                        expr))
-                          (let ([v (read log-file)])
-                            (if (eof-object? v)
-                                (error "expression result missing in log file")
-                                (let ([v (deserialize v)])
-                                  (if (gui-exn? v)
-                                      (raise (make-exn:fail
-                                              (gui-exn-message v)
-                                              (current-continuation-marks)))
-                                      v))))
-                          (error 'mreval
-                                 "expression does not match log file: ~e versus: ~e"
-                                 expr
-                                 v)))))))))))
+  (cond
+    [mred?
+     (define eh (scribble-eval-handler))
+     (define log-file (open-output-file exprs-dat-file #:exists 'truncate/replace))
+     (λ (gui-eval get-predicate? get-render get-get-width get-get-height)
+       (lambda (ev catching-exns? expr)
+         (write (serialize (if (syntax? expr)
+                               (syntax->datum expr)
+                               expr))
+                log-file)
+         (newline log-file)
+         (flush-output log-file)
+         (define result
+           (with-handlers ([exn:fail? (lambda (exn) (make-gui-exn (exn-message exn)))])
+             ;; put the call to fixup-picts in the handlers
+             ;; so that errors in the user-supplied predicates &
+             ;; conversion functions show up in the rendered output
+             (fixup-picts (get-predicate?)
+                          (get-render)
+                          (get-get-width)
+                          (get-get-height)
+                          (eh ev catching-exns? expr))))
+         (write (serialize result) log-file)
+         (newline log-file)
+         (flush-output log-file)
+         (if (gui-exn? result)
+             (raise (make-exn:fail (gui-exn-message result) (current-continuation-marks)))
+             result)))]
+    [else
+     (define log-file
+       (with-handlers ([exn:fail:filesystem? (lambda (exn) (open-input-string ""))])
+         (open-input-file exprs-dat-file)))
+     (λ (gui-eval get-predicate? get-render get-get-width get-get-height)
+       (lambda (ev catching-exns? expr)
+         (with-handlers ([exn:fail? (lambda (exn)
+                                      (when catching-exns?
+                                        (raise exn))
+                                      (void))])
+           (define v (read log-file))
+           (if (eof-object? v)
```

Comment:
Shouldn't this be converted to `cond` to allow further refactoring? Or does it hit the limit? @jackfirth

Reply:
The `let` expression in the `if` body isn't refactorable, since it has a `[v (deserialize v)]` clause.